### ISSUE TITLE
core snap is not a defined resource for charms that have other snaps as a resource definition

### ISSUE THREAD

**Issue by esunar** [esunar/test2#27]
Several charms require snaps to install their relevant software.  For an offline environment, these charms allow for a resource definition so that the local snap location can be specified.

The problem is, all snaps have a prerequesite of the core snap.

This snap is not listed as a resource for any of the current CDK charms that also have snaps as a resource.

This requires some mechanism to "sideload" the core charm.  One example of that is to use the juju cloudinit-userdata preruncmd option to wget the core snap and manually install it.

This is not an obvious solution to all end users.  There should be a core snap resource definition.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2019-05-07T14:52:55Z{.section owner}
 * owner: jhillman{.end}{.section assignee}
 * assignee: kwmonroe{.end}{.section duplicate_of}
 * duplicate of: #None{.end}{.section duplicates}
 * the following issues have been marked as duplicates of this one:{.repeated section @}
   * #{@}{.end}{.end}
 * the launchpad url was https://bugs.launchpad.net/bugs/1828063

**Comment by esunar** [comment]
(by cynerva)
Past discussion: https://github.com/stub42/layer-snap/issues/9

**Comment by esunar** [comment]
(by jhillman)
To continue that discussion here...

Stub's point isn't taking into account offline deployments.  Offline deployments aren't expecting to have automatic security updates.

And more and more deployments are offline these days, or so it seems.

**Comment by esunar** [comment]
(by jhillman)
Field critical applied, and approved by Vinod.

The workaround of sideloading the core charm is inconsistently failing.

**Comment by esunar** [comment]
(by cynerva)
One of our old scripts accomplishes this by using juju scp and juju run[1]. Can you do the same as a temporary workaround?

[1]: https://github.com/juju-solutions/cdk-shrinkwrap/blob/41cee5b906d9f8cb882dbe903ef1e8aa5bbcc962/shrinkwrap.py#L51-L55

**Comment by esunar** [comment]
(by jhillman)
If i'm reading it correctly, that script is assuming that the machines are already added, and after it installs the core snap, it then deploys the application.

I'm performing this via a bundle, using the FCE tool.  So, hence i'm using cloud-init so that once the machine is up, but before the charm software installs, the core snap is there.

So, I don't think I can use that same method, because the bundle is what is deploying the machines.

**Comment by esunar** [comment]
(by cynerva)
I'm actively working on this now.

**Comment by esunar** [comment]
(by cynerva)
PRs for etcd, kubernetes-master, kubernetes-worker:

https://github.com/charmed-kubernetes/layer-etcd/pull/153

https://github.com/charmed-kubernetes/charm-kubernetes-master/pull/19

https://github.com/charmed-kubernetes/charm-kubernetes-worker/pull/11

**Comment by esunar** [comment]
(by cynerva)
PR for vault:

https://github.com/openstack/charm-vault/pull/3

**Comment by esunar** [comment]
(by cynerva)
Gerrit code review request for vault:

https://review.opendev.org/#/c/658153/

**Comment by esunar** [comment]
(by cynerva)
Fix is available in edge channels of etcd, kubernetes-master, and kubernetes-worker with the following charm revisions:

cs:~containers/etcd-422
cs:~containers/kubernetes-master-668
cs:~containers/kubernetes-worker-529

I'll see about backporting these to stable.

**Comment by esunar** [comment]
(by cynerva)
PRs for backports to stable in etcd, kubernetes-master, kubernetes-worker:

https://github.com/charmed-kubernetes/layer-etcd/pull/155
https://github.com/charmed-kubernetes/charm-kubernetes-master/pull/20
https://github.com/charmed-kubernetes/charm-kubernetes-worker/pull/12

**Comment by esunar** [comment]
(by cynerva)
PRs for kubernetes-e2e:

master: https://github.com/charmed-kubernetes/charm-kubernetes-e2e/pull/1
stable: https://github.com/charmed-kubernetes/charm-kubernetes-e2e/pull/2

**Comment by esunar** [comment]
(by cynerva)
Fix for kubernetes-e2e is available in edge with charm revision cs:~containers/kubernetes-e2e-650

**Comment by esunar** [comment]
(by cynerva)
For etcd, kubernetes-master, kubernetes-worker, and kubernetes-e2e, we're in the process of getting a hotfix release out to stable. This is a new process for us so it's likely going to take us a few days to work out the kinks, get everything tested, and release it.

**Comment by esunar** [comment]
(by kwmonroe)
Fix for prometheus2 is available with charm revision
cs:prometheus2-9

**Comment by esunar** [comment]
(by cynerva)
We're still working on a hotfix release for etcd, kubernetes-master, kubernetes-worker, and kubernetes-e2e. Currently working through some issues with the release process. Hopefully soon we'll get a build onto the candidate channels; from there it will probably be another day or two of testing before we can release to stable.

**Comment by esunar** [comment]
(by cynerva)
Hotfix releases are currently in candidate channels with the following charm revisions:

cs:~containers/etcd-424
cs:~containers/kubernetes-master-672
cs:~containers/kubernetes-worker-531
cs:~containers/kubernetes-e2e-651

We are still working to get these tested before we can release to stable.

**Comment by esunar** [comment]
(by cynerva)
Testing of the CDK charms is currently blocked by network issues in our testing infrastructure. We are working with IS to resolve this as quickly as possible.

**Comment by esunar** [comment]
(by kwmonroe)
Fix for graylog is available with charm revision
cs:graylog-29

**Comment by esunar** [comment]
(by cynerva)
Fixes for all affected charms are now released to stable channels.

**Comment by esunar** [comment]
(by jhillman)
Field critical removed

Thanks EVERYONE for all of the effort into getting these resolved and released.

**Comment by esunar** [comment]
(by xavpaice)
Still need to have this fixed for Grafana - see https://code.launchpad.net/~graylog-charmers/charm-graylog/+git/graylog-charm/+merge/367201 for an example of the fix.

**Comment by esunar** [comment]
(by xavpaice)
Marked grafana as invalid, as the current iteration installs via .deb rather than snap.
